### PR TITLE
Creating gc command to remove unnecessary files

### PR DESCRIPTION
**Pull request by lucasncabral**
This command will remove files contained in the cache and objects directories of the ml-git metadata (.ml-git) that are not necessary for the entities that the user has in the workspace.

Usage:
```
ml-git repository gc
```

Output example:
```
  INFO - Repository: Starting the garbage collector for dataset
  INFO - Repository: Starting the garbage collector for labels
  INFO - Repository: Starting the garbage collector for model
  INFO - Repository: A total of 91200 files have been removed from C:\project\path\.ml-git
  INFO - Repository: Total reclaimed space 17.3 GB.
```

Demonstration video: https://asciinema.org/a/372547

### CODE REVIEW

File: ml_git/repository.py [hunk 1]
```diff
@@ -1035,3 +1037,57 @@ def update_entities_metadata(self):
...
+            repo_type = entity.value
+            if self.metadata_exists(repo_type):
+                log.info(output_messages['INFO_STARTING_GC'] % repo_type, class_name=REPOSITORY_CLASS_NAME)
+                any_metadata = True
+                index_path = get_index_path(self.__config, repo_type)
+                objects_path = get_objects_path(self.__config, repo_type)
+                cache_path = get_cache_path(self.__config, repo_type)
+                blobs_hashes = self._get_blobs_hashes(index_path, objects_path, repo_type)
+
+                # Clear cache
```
Comment (gilMars) [inline ml_git/repository.py:1055]: Extract method and remove the comment.

File: ml_git/repository.py [hunk 2]
```diff
@@ -1035,3 +1037,57 @@ def update_entities_metadata(self):
...
+                except Exception:
+                    log.debug(output_messages['INFO_ENTITY_DELETED'] % spec, class_name=REPOSITORY_CLASS_NAME)
+        return blobs_hashes
+
+    def check_is_valid_entity(self, repo_type, spec):
+        ref = Refs(get_refs_path(self.__config, repo_type), spec, repo_type)
+        tag, sha = ref.branch()
+        categories_path = get_path_with_categories(tag)
+        search_spec_file(repo_type, spec, categories_path)
+        return True
```
Comment (gilMars) [inline ml_git/repository.py:1093]: if the function raises an exception to validate the entity, remove the 'return True'.
change 'tag, sha = ref.branch()' -> 'tag, _ = ref.branch()'

File: ml_git/repository.py (hunk 1 above)
Comment (lucasncabral) [reply ml_git/repository.py]: Adjusted in 8d95afd.

File: ml_git/repository.py (hunk 2 above)
Comment (lucasncabral) [reply ml_git/repository.py]: Adjusted in 8d95afd.

**Review by gilMars — APPROVED** [review]

**Review by Mun1z — APPROVED** [review]

**Review by andersonpablosilva — APPROVED** [review]

File: ml_git/file_system/objects.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    def garbage_collector(self, blobs_hashes):
+        used_blobs = self._get_used_blobs(blobs_hashes)
+        count_removed_objects, reclaimed_objects_space = remove_unnecessary_files(used_blobs,
+                                                                                  os.path.join(self._objects_path, HASH_FS_CLASS_NAME.lower()))
+        log.debug(output_messages['INFO_REMOVED_FILES'] % (number_to_human_format(count_removed_objects),
+                                                           self._objects_path))
+        return count_removed_objects, reclaimed_objects_space
+
+    def _get_used_blobs(self, descriptor_hashes):
```
Comment (diego-pedroso) [inline ml_git/file_system/objects.py:54]: Please define the function before its usage.

File: ml_git/repository.py [hunk 4]
```diff
@@ -1060,3 +1061,50 @@ def update_entities_metadata(self):
...
+        if not any_metadata:
+            log.error(output_messages['ERROR_UNINITIALIZED_METADATA'], class_name=REPOSITORY_CLASS_NAME)
+            return
+        log.info(output_messages['INFO_REMOVED_FILES'] % (number_to_human_format(removed_files),
+                                                          os.path.join(get_root_path(), '.ml-git')),
+                 class_name=REPOSITORY_CLASS_NAME)
+        log.info(output_messages['INFO_RECLAIMED_SPACE'] % size(reclaimed_space, system=alternative),
+                 class_name=REPOSITORY_CLASS_NAME)
+
+    def _get_blobs_hashes(self, index_path, objects_path, repo_type):
```
Comment (diego-pedroso) [inline ml_git/repository.py:1094]: Please define the function before its usage.

File: ml_git/repository.py [hunk 5]
```diff
@@ -1060,3 +1061,50 @@ def update_entities_metadata(self):
...
+            for spec in dirs:
+                try:
+                    self.check_is_valid_entity(repo_type, spec)
+                    idx = MultihashIndex(spec, index_path, objects_path)
+                    blobs_hashes.extend(idx.get_hashes_list())
+                except Exception:
+                    log.debug(output_messages['INFO_ENTITY_DELETED'] % spec, class_name=REPOSITORY_CLASS_NAME)
+        return blobs_hashes
+
+    def check_is_valid_entity(self, repo_type, spec):
```
Comment (diego-pedroso) [inline ml_git/repository.py:1106]: Please define the function before its usage and add a leading underscore.

File: ml_git/utils.py [hunk 6]
```diff
@@ -237,3 +239,43 @@ def group_files_by_path(files):
...
+            if file not in filenames:
+                file_path = os.path.join(root, file)
+                reclaimed_space += Path(file_path).stat().st_size
+                set_write_read(file_path)
+                os.unlink(file_path)
+                count += 1
+    return reclaimed_space, count
+
+
+def number_to_human_format(num):
```
Comment (diego-pedroso) [inline ml_git/utils.py:275]: Have we considered using https://github.com/jmoiron/humanize?

File: tests/integration/test_35_gc.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        metadata_path = os.path.join(self.tmp_dir, ML_GIT_DIR)
+        number_of_files = 0
+        original_size = 0
+        for root, dirs, files in os.walk(metadata_path):
+            for file in files:
+                number_of_files += 1
+                original_size += Path(os.path.join(root, file)).stat().st_size
+        return original_size, number_of_files
+
+    def check_result(self, result, entity, original_size, number_of_files, expected_removed_files, expected_reclaimed_space):
```
Comment (diego-pedroso) [inline tests/integration/test_35_gc.py:43]: Please add a leading underscore.

File: ml_git/file_system/objects.py (hunk 3 above)
Comment (lucasncabral) [reply ml_git/file_system/objects.py:47]: Done in baaa703

File: ml_git/repository.py (hunk 4 above)
Comment (lucasncabral) [reply ml_git/repository.py]: Done in baaa703

File: ml_git/repository.py (hunk 5 above)
Comment (lucasncabral) [reply ml_git/repository.py]: Done in baaa703

File: ml_git/utils.py (hunk 6 above)
Comment (lucasncabral) [reply ml_git/utils.py]: Taking into account only the context of this problem, it was seen that the use of a library is an unnecessary effort to solve a relatively small problem.

However, in other parts of the code we use [hurry.filesize](https://pypi.org/project/hurry.filesize/) to transform human-readable file sizes.  As [humanize](https://pypi.org/project/humanize/) also has a similar implementation, it becomes more interesting to use this lib for this point and the others, in order to standardize this type of treatment in ml-git.

This change was made in baaa703

File: tests/integration/test_35_gc.py (hunk 7 above)
Comment (lucasncabral) [reply tests/integration/test_35_gc.py]: Done in baaa703

**Review by diego-pedroso — APPROVED** [review]

**Pull request merged**